### PR TITLE
zig info command - print lib path, std path and global cache path

### PR DESCRIPTION
**Pull request by Sergeeeek**
Implements the enhancement #4427

Creating this PR before everything is finished to get early feedback.

A couple of questions:
1. ~~stage0 uses Blake2 to compute the compiler id hash, can I use that for self hosted or does it need to be re-implemented?~~ Will be implemented in #4635 
2. ~~How to pass a version string from build.zig? Are there any examples/docs? I haven't found anything.~~ Will use addBuildOption

**Progress**
---
**Self hosted**:
- [x] Add an `info` command to zig
- [x] lib path
- [x] std path
- [x] global cache path (different for self hosted and stage1)
- [x] zig info --json
- [x] zig info --help

Example output:
```sh
$ zig info
lib_dir         	/home/sergeeeek/Work/zig/build/lib/zig
std_dir         	/home/sergeeeek/Work/zig/build/lib/zig/std
global_cache_dir	/home/sergeeeek/.cache/zig/(stage1|self_hosted)

$ zig info --json
{
  "lib_dir": "/home/sergeeeek/Work/zig/build/lib/zig",
  "std_dir": "/home/sergeeeek/Work/zig/build/lib/zig/std",
  "global_cache_dir": "/home/sergeeeek/.cache/zig/(stage1|self_hosted)"
}

$ zig info --help
Usage: zig info [options]

   Outputs path to zig lib dir, std dir and the global cache dir.

Options:
   --help                 Print this help and exit
   --json                 Output as json instead of a table format
```

**Stage 1**:
- [x] Add an `info` command to zig
- [x] Call the self-hosted info command, so the output is the same

### CODE REVIEW

**Comment by Sergeeeek** [comment]
> How to pass a version string from build.zig? Are there any examples/docs? I haven't found anything.

I'm thinking creating a separate file called `version.txt` at the root of the project and reading it from cmake and at comptime in self hosted.

I guess I'd also need to call `git` to get the revision inside a comptime, not sure if that's a good solution though.

Any help on choosing the best approach would be appreciated.

**Comment by Sergeeeek** [comment]
Found the zig blake2 implementation, working on translating cache_hash.cpp into zig, which is used by current `get_compiler_id`

**Comment by andrewrk** [comment]
> Found the zig blake2 implementation, working on translating cache_hash.cpp into zig, which is used by current `get_compiler_id`

I'll work on merging #4635 soon, I think you might be duplicating efforts here

**Comment by ifreund** [comment]
I'd argue that json output is overkill for this. A simple table would be easier to work with from shell scripts without depending on something like `jq`.

**Comment by data-man** [comment]
> A simple table would be easier to work

Or to add additional argument:
```zig info [--json]```

**Comment by Sergeeeek** [comment]
> > Found the zig blake2 implementation, working on translating cache_hash.cpp into zig, which is used by current `get_compiler_id`
> 
> I'll work on merging #4635 soon, I think you might be duplicating efforts here

Oh didn't notice that PR, I'll stop then. Didn't get that far anyway :)

> > A simple table would be easier to work
> 
> Or to add additional argument:
> `zig info [--json]`

Agreed, that's a good idea

**Comment by Sergeeeek** [comment]
Figuring out how to pass version from `build.zig` takes too long for me :sweat_smile:

I'd like to split self hosted version resolution and the `--json` argument into a separate PR and merge this in its current state.

File: deps/SoftFloat-3e/COPYING.txt [hunk 1]
```diff
@@ -1,37 +1,37 @@
...
+EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE, ARE
+DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE FOR ANY
+DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
```
Comment (alexnask) [inline deps/SoftFloat-3e/COPYING.txt:37]: This happens to me too somethimes >.>

Comment (daurnimator) [reply deps/SoftFloat-3e/COPYING.txt:37]: same :(

lets just change the line endings and be done with it

Comment (Sergeeeek) [reply deps/SoftFloat-3e/COPYING.txt:37]: Should I remove that from this PR?

Comment (daurnimator) [reply deps/SoftFloat-3e/COPYING.txt:37]: See #5489

> Should I remove that from this PR?

yes.

Comment (Sergeeeek) [reply deps/SoftFloat-3e/COPYING.txt]: Removed :)

File: src-self-hosted/print_info.zig [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+            try stdout.writeAll(usage_info);
+            return;
+        } else {
+            std.debug.warn("Unknown argument passed to info command: {}\n", .{arg});
+            process.exit(1);
+        }
+    }
+
+    if (json_format) {
+        try info.toJSON(bos_stream);
```
Comment (daurnimator) [inline src-self-hosted/print_info.zig:124]: Just call `try json.stringify(info, .{.whitespace = .{.indent = .{.Space = 2}}}, bos_stream);` here?

File: src-self-hosted/print_info.zig [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+    var info = try CompilerInfo.init(allocator, compiler_type);
+    defer info.deinit(allocator);
+
+    var bos = io.bufferedOutStream(stdout);
+    const bos_stream = bos.outStream();
+
+    var json_format = false;
+    for (cmd_args) |arg| {
+        if (mem.eql(u8, arg, "--json")) {
+            json_format = true;
```
Comment (daurnimator) [inline src-self-hosted/print_info.zig:113]: Maybe also have a `--text` option to set it back to non-json?

Perhaps it should be a `--format=json` argument instead?

Comment (Sergeeeek) [reply src-self-hosted/print_info.zig:113]: It's text by default, why add an additional `--text` argument?

**Comment by Sergeeeek** [comment]
Found an inconsistency between Linux and OSX that affects this PR: `fs.selfExePath()` on Linux resolves symlinks and OSX it doesn't.

The fix to that is very simple, I can just call `os.readlinkZ` on OSX like it currently does on linux. Can I include that in this PR?

File: src-self-hosted/print_info.zig (hunk 2 above)
Comment (Sergeeeek) [reply src-self-hosted/print_info.zig]: Fixed

**Comment by Sergeeeek** [comment]
Actually, looks like symlinks in `selfExePath` are *only* resolved on Linux. Not sure what to do about it, but I would propose to resolve them on all platforms.

For example `brew` usually installs binaries into it's own location and then links them to `/usr/local/bin`, in which case `zig info` would stop working.

File: src-self-hosted/print_info.zig (hunk 3 above)
Comment (daurnimator) [reply src-self-hosted/print_info.zig:103]: It's common in many command line tools so that aliases and defaults set in wrappers can be undone.

**Comment by Sergeeeek** [comment]
Another solution is to remove link resolution in `selfExePath` on linux and do it inside zig info, but I'm not sure how much stuff is dependent on that behavior

File: src-self-hosted/print_info.zig (hunk 3 above)
Comment (Sergeeeek) [reply src-self-hosted/print_info.zig]: Added a `--format` option, also changed error messages to better match other zig commands.

**Comment by andrewrk** [comment]
Thanks @Sergeeeek. I made some adjustments and then it landed in e26dda5308d5c853d195b65558454db4058ff218.

**Pull request closed without merging**